README.md
This is the README file for Domain-Robot, an approach
for a unified domain registration API in perl.

## Description

Domain-Robot will provide a unified API for implementing
plugins that interact with any domain registrar which
offers some kind of API.

## Installation

This package uses Dist::Zilla.

Use

dzil build

to create a release tarball which can be
unpacked and installed like any other EUMM
distribution.

perl Makefile.PL

make

make test

make install

## Documentation

Please see perldoc Doamin::Robot.


Changes
Revision history for Domain-Robot

0.01    2012-12-26
        First version, released on an unsuspecting world.



lib/Domain/Registrar.pm
package Domain::Robot::Registrar;
# ABSTRACT: baseclass for any registrar plugin

use 5.010_000;
use mro 'c3';
use feature ':5.10';

use Moose;
use namespace::autoclean;

# use IO::Handle;
# use autodie;
# use MooseX::Params::Validate;
# use Carp;
# use English qw( -no_match_vars );
# use Try::Tiny;
use LWP::UserAgent;

# extends ...
# has ...
has 'parent' => (
    'is'    => 'ro',
    'isa'   => 'Domain::Robot',
    'required' => 1,
);

has 'priority' => (
    'is'    => 'ro',
    'isa'   => 'Int',
    'lazy'  => 1,
    'builder' => '_init_priority',
);

has '_ua' => (
    'is'      => 'rw',
    'isa'     => 'LWP::UserAgent',
    'lazy'    => 1,
    'builder' => '_init_ua',
);

has 'username' => (
    'is'       => 'rw',
    'isa'      => 'Str',
    'required' => 1,
);

has 'base_url' => (
    'is'    => 'rw',
    'isa'   => 'Str',
    'lazy'  => 1,
    'builder' => '_init_base_url',
);

has 'password' => (
    'is'       => 'rw',
    'isa'      => 'Str',
    'required' => 1,
);

has 'name' => (
    'is'      => 'rw',
    'isa'     => 'Str',
    'default' => 'Unnamed Robot',
);

has 'description' => (
    'is'      => 'rw',
    'isa'     => 'Str',
    'default' => 'Unnamed Robot',
);
# with ...
with qw(Log::Tree::RequiredLogger);
# initializers ...
sub _init_priority { return 0; }

sub _init_ua {
    my $self = shift;

    my $UA = LWP::UserAgent::->new();
    $UA->agent('Domain::Robot/1.0');

    return $UA;
}

# your code here ...

no Moose;
__PACKAGE__->meta->make_immutable;

1;

__END__

=head1 SYNOPSIS

    use Domain::Robot::Registrar;
    my $Mod = Domain::Robot::Registrar::->new();

=head1 DESCRIPTION

Some description.

=cut


lib/Domain/Registrar/RRPProxy.pm
package Domain::Robot::Registrar::RRPProxy;
# ABSTRACT: client for the RRPProxy API

use 5.010_000;
use mro 'c3';
use feature ':5.10';

use Moose;
use namespace::autoclean;

# use IO::Handle;
# use autodie;
# use MooseX::Params::Validate;

our $VERSION = '@VERSION@';

use XMLRPC::Lite;
use Try::Tiny;
use POSIX;    # for mktime

extends 'Domain::Robot::Registrar';

has 'base_url' => (
    'is'       => 'ro',
    'isa'      => 'Str',
    'required' => 1,
);

has 'client' => (
    'is'      => 'ro',
    'isa'     => 'XMLRPC::Lite',
    'lazy'    => 1,
    'builder' => '_init_client',
);

sub _init_client {
    my $self = shift;

    my $Client = XMLRPC::Lite->proxy( $self->base_url() );
    try {
        $Client->transport()->ssl_opts( verify_hostname => 0, );
    };

    return $Client;
}

# call('Api.xcall', (s_login => login, s_pw => pass, command => cmd, domain => ...));
sub _api_call {
    my $self     = shift;
    my $opts_ref = shift;
    my $key_col  = shift || undef;

    $opts_ref->{'s_login'} = $self->username();
    $opts_ref->{'s_pw'}    = $self->password();

    my $resp = $self->client()->call( 'Api.xcall', $opts_ref )->result();
    my $result_ref = {};

    # if key_col is specified the processing mode is changed
    # then not everything is copied but only the information from
    # the respective row ... this is dirty but the way this api works :(
    if ( $key_col && $resp->{'PROPERTY'}->{$key_col} ) {
        foreach my $i ( 0 .. $#{ $resp->{'PROPERTY'}->{$key_col} } ) {
            my $key = $resp->{'PROPERTY'}->{$key_col}->[$i];
            foreach my $prop ( keys %{ $resp->{'PROPERTY'} } ) {
                if ( scalar( @{ $resp->{'PROPERTY'}->{$prop} } ) == 1 ) {
                    $result_ref->{$key}->{ 'rrpproxy::' . $prop } = $resp->{'PROPERTY'}->{$prop}->[0];
                }
                else {
                    $result_ref->{$key}->{ 'rrpproxy::' . $prop } = $resp->{'PROPERTY'}->{$prop}->[$i];
                }
            }
        }
    }
    else {
        foreach my $prop ( keys %{ $resp->{'PROPERTY'} } ) {
            if ( scalar( @{ $resp->{'PROPERTY'}->{$prop} } ) == 1 ) {
                $result_ref->{ 'rrpproxy::' . $prop } = $resp->{'PROPERTY'}->{$prop}->[0];
            }
            elsif ( scalar( @{ $resp->{'PROPERTY'}->{$prop} } ) > 1 ) {
                $result_ref->{ 'rrpproxy::' . $prop } = $resp->{'PROPERTY'}->{$prop};
            }
        }
    }

    return $result_ref;
}

# Action 1.1 AddDomain
sub domain_create {
    my $self   = shift;
    my $domain = shift;
    my $opts   = shift;

    my $cmd_ref = {};
    $cmd_ref->{'command'}         = 'AddDomain';
    $cmd_ref->{'domain'}          = $domain;
    $cmd_ref->{'ownercontact0'}   = $opts->{'ownerc'};
    $cmd_ref->{'admincontact0'}   = $opts->{'adminc'};
    $cmd_ref->{'techcontact0'}    = $opts->{'techc'};
    $cmd_ref->{'billingcontact0'} = $opts->{'billingc'};

    my $result = $self->_api_call($cmd_ref);
    print Dumper($result);

    # TODO implement
    return;
}

# Action 1.2 CheckDomain
sub domain_check {
    my $self   = shift;
    my $domain = shift;

    my $cmd_ref = {};
    $cmd_ref->{'command'} = 'CheckDomain';
    $cmd_ref->{'domain'}  = $domain;
    my $result_ref = $self->_api_call($cmd_ref);
    print Dumper($result_ref);

    # TODO implement
    return;
}

# Action 1.3 CheckDomains
# Action 1.4 StatusDomain
sub domain_info {
    my $self   = shift;
    my $domain = shift;

    my $result_ref = $self->_api_call( { 'command' => 'StatusDomain', 'domain' => $domain, } );
    return $result_ref;
}

# Action 1.5 ModifyDomain
# Action 1.6 DeleteDomain
# Action 1.7 RenewDomain
# Action 1.8 PushDomain
# Action 1.9 TradeDomain
# Action 1.10 GetZone
# Action 1.11 QueryDomainList
sub domain_list {
    my $self = shift;

    # TODO parse known fields into default namespace, e.g. inwx::crDate -> create_ts (unix ts)

    my $result_ref = $self->_api_call( { 'command' => 'QueryDomainList', 'wide' => 1, }, 'DOMAIN', );
    my %dates = (
        'DOMAIN_RENEWAL_DATE'                 => 'renewal_ts',
        'DOMAIN_UPDATED_DATE'                 => 'update_ts',
        'DOMAIN_CREATED_DATE'                 => 'create_ts',
        'DOMAIN_REGISTRATION_EXPIRATION_DATE' => 'expiration_ts',
    );
    foreach my $domain ( keys %{$result_ref} ) {
        foreach my $skey ( keys %dates ) {
            my $dkey = $dates{$skey};
            $result_ref->{$dkey} = $self->_parse_datetime( $result_ref->{ 'rrpproxy::' . $skey } );
        }

        # TODO get additional domain information via StatusDomain?
    }

    return $result_ref;
}

sub _parse_datetime {
    my $self        = shift;
    my $date_string = shift;

    # YYYY-MM-DD HH:mm:ss
    if ( $date_string && $date_string =~ m/(\d{4})-(\d{2})-(\d{2}) (\d{2}):(\d{2}):(\d{2})/ ) {
        my ( $year, $month, $dom, $hour, $minute, $second ) = ( $1, $2, $3, $4, $5, $6 );
        my $ts = POSIX::mktime( $second, $minute, $hour, $dom, $month - 1, $year - 1900 );
        return $ts;
    }
    else {
        return 0;
    }
}

# Action 1.12 QueryDomainListByContact
# Action 1.13 GetZoneInfo
# Action 2.1 AddDomainApplication
# Action 2.2 CheckDomainApplication
# Action 2.3 StatusDomainApplication
# Action 2.4 DeleteDomainApplication
# Action 2.5 QueryDomainApplicationList
# Action 3.1 TransferDomain
# Action 3.2 CheckDomainTransfer
# Action 3.3 StatusDomainTransfer
# Action 3.4 SetAuthcode
# Action 3.5 ActivateTransfer
# Action 3.6 QueryTransferList
# Action 4.1 AddNameserver
# Action 4.2 CheckNameserver
# Action 4.3 StatusNameserver
# Action 4.4 ModifyNameserver
# Action 4.5 DeleteNameserver
# Action 4.6 QueryNameserverList
# Action 5.1 AddContact
sub contact_create {
    my $self = shift;
    my $opts = shift || {};

    my $cmd_ref = {};
    $cmd_ref->{'command'} = 'AddContact';
    foreach my $key (qw(title firstname lastname organization street zip city country phone fax email)) {
        if ( !$opts->{$key} ) {
            warn "Missing option to contact_create: $key!\n";
        }
        $cmd_ref->{$key} = $opts->{$key};
    }

    my $resp = $self->_api_call($cmd_ref);
    return $resp->{'rrpproxy::CONTACT'};
}

# Action 5.2 StatusContact
# Action 5.3 ModifyContact
# Action 5.4 DeleteContact
# Action 5.5 CheckContact
# Action 5.6 CheckContacts
# Action 5.7 CloneContact
# Action 5.8 QueryContactList
sub contact_list {
    my $self = shift;

    my $cmd_ref = {};
    $cmd_ref->{'command'} = 'QueryContactList';
    $cmd_ref->{'wide'}    = 1;

    my $result_ref = $self->_api_call( $cmd_ref, 'CONTACT', );
    return $result_ref;
}

# Action 6.1 QueryObjectLogList
# Action 6.2 StatusObjectLog
# Action 7.1 QueryEventList
# Action 7.2 StatusEvent
# Action 7.3 DeletEvent
# Action 8.1 StatusRegistrar
# Action 8.2 ModifyRegistrar
# Action 8.3 StatusAccount
# Action 8.4 GetAppendix
# Action 8.5 ActiveAppendix
# Action 8.6 QueryAppendixList
# Action 8.7 QueryRegistrarStatisticList
# Action 9.1 ConvertCurrency
# Action 9.2 ChargeCreditcard
# Action 9.3 QueryExchangeRates
sub accounting_exchangerates {
    my $self = shift;

    return $self->_api_call( { 'command' => 'QueryExchangeRates', } );
}

# Action 9.4 QueryAccountingList
sub accounting_log {
    my $self = shift;

    return $self->_api_call( { 'command' => 'QueryAccountingList', } );
}

# Action 9.5 QueryZoneList
sub domain_getprices {
    my $self = shift;

    my $domains = $self->_api_call( { 'command' => 'QueryZoneList', } );

    my $result_ref = {};
    foreach my $i ( 1 .. $#{ $domains->{'PROPERTY'}->{'ZONE'} } ) {
        my $key = $domains->{'PROPERTY'}->{'ZONE'}->[$i];
        foreach my $prop (qw(3RDS RESTORE SETUP APPLICATION CURRENCY TRADE DOMAIN_COUNT ANNUAL TRANSFER ACTIVE PERIODTYPE)) {
            $result_ref->{$key}->{ 'rrpproxy::' . $prop } = $domains->{'PROPERTY'}->{$prop}->[$i];
        }
    }

    # parse known fields into default namespace, e.g. inwx::crDate -> create_ts (unix ts)
    my %mapping = (
        'TRADE'       => 'price_trade',
        'RESTORE'     => 'price_restore',
        'APPLICATION' => 'price_application',
        'TRANSFER'    => 'price_transfer',
        'ANNUAL'      => 'price_renew',
    );
    my @periods = qw(create renew transfer trade);
    foreach my $domain ( keys %{$result_ref} ) {
        if ( $result_ref->{$domain}->{'rrpproxy::PERIODTYPE'} ne 'YEAR' ) {
            delete $result_ref->{$domain};
            next;
        }
        $result_ref->{$domain}->{'price_create'} =
          sprintf( "%.2f", $result_ref->{$domain}->{'rrpproxy::SETUP'} + $result_ref->{$domain}->{'rrpproxy::ANNUAL'} );
        foreach my $skey ( keys %mapping ) {
            my $dkey = $mapping{$skey};
            $result_ref->{$domain}->{$dkey} = sprintf( "%.2f", $result_ref->{$domain}->{ 'rrpproxy::' . $skey } );
        }
        foreach my $period (@periods) {
            $result_ref->{$domain}->{ 'period_' . $period } = 1;
        }
        $result_ref->{$domain}->{'currency'} = $result_ref->{$domain}->{'rrpproxy::CURRENCY'};
    }

    return $result_ref;
}

# Action 9.6 QueryServiceList
# Action 9.7 DomainPrice
# Action 9.8 ServicePrice

no Moose;
__PACKAGE__->meta->make_immutable;

1;

__END__

=head1 SYNOPSIS

    use Domain::Robot::RRPProxy;
    use Data::Dumper;
    
    my $r = Domain::Robot::RRPProxy::->new({
      'logger' => $l,  
      'username' => 'username',
      'password' => 'password',
      'base_url' => 'https://api-ote.rrpproxy.net:8083/xmlrpc/',
    });
    print Dumper($r->domain_list());
    
=cut


lib/Domain/Registrar/Schlundtech.pm
package Domain::Robot::Registrar::Schlundtech;
# ABSTRACT: client for the Schlundtech API
# Note: should work for InternetX, too

use 5.010_000;
use mro 'c3';
use feature ':5.10';

use Moose;
use namespace::autoclean;

# use IO::Handle;
# use autodie;
# use MooseX::Params::Validate;

use XML::Simple;
use LWP::UserAgent;

extends 'Domain::Robot::Registrar';

has 'email_contact' => (
    'is'       => 'rw',
    'isa'      => 'Str',
    'required' => 0,
);

has 'email_robot' => (
    'is'       => 'rw',
    'isa'      => 'Str',
    'required' => 0,
);

sub _init_base_url {
    return 'https://gateway.schlundtech.de';
}

sub _api_call {
    my $self = shift;
    my $task = shift;
    my $opts = shift || {};

    my $request = {};
    $request->{'auth'}->{'user'}     = $self->username();
    $request->{'auth'}->{'password'} = $self->password();
    $request->{'auth'}->{'context'}  = 10;                  # always 10
    $request->{'language'}           = 'en';                # de or en
    $request->{'task'}               = $task;

    my $xml_str = XMLout($request, RootName => 'request', XMLDecl => 1, NoAttr => 1, );

    my $req = HTTP::Request::->new( POST => $self->base_url(), );
    $req->content_type('text/xml');
    $req->content($xml_str);
    my $res = $self->_ua()->request($req);

    if ( $res->is_success() ) {
        my $result_ref = XMLin( $res->content() );
        return $result_ref;
    }
    else {
        return;
    }
}

# 1.x - no methods
# 2.1 Domain Create - 0101
sub domain_create {
    my $self   = shift;
    my $domain = shift;
    my $ownerc = shift;
    my $nserver1 = shift;
    my $nserver2 = shift;
    my $ip = shift;

    my $task = {};
    $task->{'code'} = '0101';
    $task->{'domain'}->{'confirm_order'} = 1;
    $task->{'domain'}->{'name'} = $domain;
    $task->{'domain'}->{'ownerc'} = $ownerc;
    push(@{$task->{'domain'}->{'nserver'}},{
        'name'  => $nserver1,
    });
    push(@{$task->{'domain'}->{'nserver'}},{
        'name'  => $nserver2,
    });
    $task->{'domain'}->{'zone'}->{'ip'} = $ip;
    $task->{'domain'}->{'zone'}->{'ns_action'} = 'complete';
    
    use Data::Dumper;
    print Dumper($task);
    
    #my $resp = $self->_api_call($task);
    my $resp;
    return $resp;
}

# 2.2 Domain Create Plus - no code/no method, see 2.1
# 2.3 Domain Update - 0102
sub domain_update {
    my $self   = shift;
    my $domain = shift;

    my $task = {};
    $task->{'code'} = '0102';
    $task->{'domain'}->{'name'} = $domain;

    # TODO set options to be updated
    my $resp = $self->_api_call($task);
    
    return $resp;
}

# 2.4 Domain Update Plus - no code/no method, see 2.3
# 2.5 Domain Ownerchange - 0104010
sub domain_trade {
    my $self   = shift;
    my $domain = shift;

    my $task = {};
    $task->{'code'} = '0104010';
    $task->{'domain'}->{'name'} = $domain;

    # TODO disabled
    return;
}

# 2.6 Domain Delete - 0103
sub domain_delete {
    my $self   = shift;
    my $domain = shift;

    my $task = {};
    $task->{'code'} = '0103';
    $task->{'domain'}->{'name'} = $domain;

    # TODO disabled ...
    return;
}

# 2.7 Domain Inquire - 0105
sub domain_info {
    my $self   = shift;
    my $domain = shift;

    my $task = {};
    $task->{'code'}             = '0105';
    $task->{'domain'}->{'name'} = $domain;
    $task->{'key'}              = 'payable';

    my $resp = $self->_api_call($task);
    
    return $resp;
}

sub domain_list {
    my $self = shift;
    my $opts = shift || {};

    my $task = {};
    $task->{'code'}               = '0105';
    $task->{'view'}->{'offset'}   = 0;
    $task->{'view'}->{'limit'}    = 10_000;
    $task->{'view'}->{'children'} = 1;
    $task->{'view'}->{'show_handle_details'} = 'ownerc,adminc,techc,zonec';

    my $resp = $self->_api_call($task);
    
    if($resp->{'result'}->{'status'}->{'type'} eq 'success') {
        return $resp->{'result'}->{'data'}->{'domain'};
    } else {
        return;
    }
}

# 2.8 Domain Status - 0102002
sub domain_status {
    my $self   = shift;
    my $domain = shift;
    my $status = shift; # lock,hold,hold-lock,active

    my $task = {};
    $task->{'code'} = '0102002';
    $task->{'domain'}->{'name'} = $domain;
    $task->{'domain'}->{'registry_status'} = $status;

    my $resp = $self->_api_call($task);

    return $resp;
}

# 3.1 - no method
# 3.2 Cancelation Create - 0103101
sub cancelation_create {
    my $self   = shift;
    my $domain = shift;
    my $type   = shift;

    my $task = {};
    $task->{'code'}                    = '0103101';
    $task->{'cancelation'}->{'domain'} = $domain;
    $task->{'cancelation'}->{'type'}   = $type;

    # TODO add opts

    my $resp = $self->_api_call($task);
    return $resp;
}

# 3.3 Cancelation Update - 0103102
sub cancelation_update {
    my $self   = shift;
    my $domain = shift;
    my $type   = shift;

    my $task = {};
    $task->{'code'}                    = '0103102';
    $task->{'cancelation'}->{'domain'} = $domain;
    $task->{'cancelation'}->{'type'}   = $type;

    # TODO add opts

    my $resp = $self->_api_call($task);
    return $resp;
}

# 3.4 Cancelation Delete - 0103103
sub cancelation_delete {
    my $self   = shift;
    my $domain = shift;

    my $task = {};
    $task->{'code'} = '0103103';
    $task->{'cancelation'}->{'domain'} = $domain;

    # TODO add opts

    my $resp = $self->_api_call($task);
    return $resp;
}

# 3.5 Cancelation Inquire (single) - 0103104
sub cancelation_list {
    my $self = shift;

    my $task = {};
    $task->{'code'} = '0103104';

    # TODO add opts

    my $resp = $self->_api_call($task);
    return $resp;
}

# 3.6 Cancelation Inquire (multi) - see 3.5
# TODO append this information to domain_list?
# 4.1 Domain Transfer In - 0104
sub domain_transfer {
    my $self   = shift;
    my $domain = shift;

    my $task = {};
    $task->{'code'} = '0104';
    $task->{'domain'}->{'name'} = $domain;

    # TODO add opts

    my $resp = $self->_api_call($task);
    return $resp;
}

# 4.2 Domain Status - 0102002
# see 2.8
# 4.3 AuthInfo1 Create - 0113001
sub domain_authinfo1create {
    my $self   = shift;
    my $domain = shift;

    my $task = {};
    $task->{'code'} = '0113001';
    $task->{'domain'}->{'name'} = $domain;

    my $resp = $self->_api_call($task);
    return $resp;
}
# 4.4 AuthInfo1 Delete - 0113002
sub domain_authinfo1delete {
    my $self   = shift;
    my $domain = shift;

    my $task = {};
    $task->{'code'} = '0113002';
    $task->{'domain'}->{'name'} = $domain;

    my $resp = $self->_api_call($task);
    return $resp;
}
# 5.1 - no method
# 5.2 Zone Create - 0201
sub nameserver_create {
    my $self = shift;
    my $name = shift;
    my $ip = shift;

    my $task = {};
    $task->{'code'} = '0201';
    $task->{'zone'}->{'name'} = $name;
    $task->{'zone'}->{'ns_action'} = 'complete';
    $task->{'zone'}->{'main'}->{'value'} = $ip;

    # TODO add opts

    my $resp = $self->_api_call($task);
    return $resp;
}

# 5.3 Zone Update - 0202
sub nameserver_update {
    my $self = shift;

    my $task = {};
    $task->{'code'} = '0202';

    # TODO add opts

    my $resp = $self->_api_call($task);
    return $resp;
}

# 5.4 Zone Delete - 0203
sub nameserver_delete {
    my $self = shift;

    my $task = {};
    $task->{'code'} = '0203';

    # TODO add opts

    my $resp = $self->_api_call($task);
    return $resp;
}

# 5.5 Zone Import - 0204
sub nameserver_import {
    my $self = shift;

    my $task = {};
    $task->{'code'} = '0204';

    # TODO add opts

    my $resp = $self->_api_call($task);
    return $resp;
}
# 5.6 Zone Inquire - 0205
sub nameserver_list {
    my $self = shift;
    my $extended = shift || 0;

    my $task = {};
    $task->{'code'} = '0205';
    $task->{'view'}->{'offset'} = 0;
    $task->{'view'}->{'limit'} = 1000;
    $task->{'view'}->{'children'} = 1;

    my $resp = $self->_api_call($task);
    
    if($extended) {
        foreach my $zone (keys %{$resp->{'result'}->{'data'}->{'zone'}}) {
            my $zone_data = $self->nameserverset_list($zone);
            $resp->{'result'}->{'data'}->{'zone'}->{$zone}->{'zone'} = $zone_data->{'result'}->{'data'}->{'zone'};
        }
    }
    
    if($resp->{'result'}->{'status'}->{'type'} eq 'success') {
        return $resp->{'result'}->{'data'};
    } else {
        # TODO log error
        return;
    }
}

# Action 10.5 - 0205
sub nameserverset_list {
    my $self = shift;
    my $zone = shift;
    
    my $task = {};
    $task->{'code'} = '0205';
    $task->{'zone'}->{'name'} = $zone;
    
    my $resp = $self->_api_call($task);
    return $resp;
}

# 6.1 - no method
# 6.2 Handle Create - 0301
sub contact_create {
    my $self = shift;

    my $task = {};
    $task->{'code'} = '0301';

    # TODO add opts

    my $resp = $self->_api_call($task);
    return $resp;
}

# 6.3 Handle Update - 0302
sub contact_update {
    my $self = shift;

    my $task = {};
    $task->{'code'} = '0302';

    # TODO add opts

    my $resp = $self->_api_call($task);
    return $resp;
}

# 6.4 Handle Delete - 0303
sub contact_delete {
    my $self = shift;

    my $task = {};
    $task->{'code'} = '0303';

    # TODO add opts

    my $resp = $self->_api_call($task);
    return $resp;
}

# 6.5 Handle Inquire - 0304
sub contact_list {
    my $self = shift;

    my $task = {};
    $task->{'code'} = '0304';

    # TODO add opts

    my $resp = $self->_api_call($task);
    return $resp;
}

# 6.6 - no method

# api provides no way of retreiving the pricelist
# you need to manage the pricelist for this provider yourself
sub domain_getprices {
    my $self = shift;

    return;
}

no Moose;
__PACKAGE__->meta->make_immutable;

1;

__END__
